### PR TITLE
[BB-2959] Mcka User progress migration endpoint and background task

### PR DESCRIPTION
**Pull request by 0x29a**
This PR adds endpoint and background task, that can be used to migrate course progress of some user (enrollment, block completions, module states, submissions). 

**Testing instructions**:

1. Run following script. Follow comments.
    ```python
    #!/usr/bin/python3

    import csv
    import requests

    from io import StringIO


    ### Prepare csv with data of users to be migrated


    # You have to create course and two users.
    # Make sure you have toggled ENABLE_COMPLETION_TRACKING waffle switch. You can do this in admin panel.
    # After that, create some blocks in course, enroll source user in this course and submit answers for blocks.
    row = ('course-v1:OpenCraft+Course1+C1', 'source@example.com', 'target@example.com', '')

    file = StringIO()
    writer = csv.writer(file)
    csv_header = ('course', 'source_email', 'dest_email', 'outcome')
    writer.writerow(csv_header)
    writer.writerow(row)

    file.seek(0)

    ### Send request to migrate user progress

    migrate_url = 'http://edx.devstack.lms:18000/api/user/v1/progress/migrate/'

    # Relogin in LMS, make any request to LMS, then open browser debugger, and retreive these
    # values from request's `Cookie` header
    jwt_header = 'eyJhbGciOiJSUzUxMiIsImtpZCI6ImRldnN0YWNrX2tleSJ9'
    jwt_payload = 'eyJzY29wZXMiOiBbImVtYWlsIiwgInByb2ZpbGUiXSwgImFkbWluaXN0cmF0b3IiOiB0cnVlLCAic3VwZXJ1c2VyIjogdHJ1ZSwgInByZWZlcnJlZF91c2VybmFtZSI6ICJlZHgiLCAiZmlsdGVycyI6IFsidXNlcjptZSJdLCAiaWF0IjogMTYwMzk5MTAzMywgInN1YiI6ICI1MGNmOTJmOWEzZDg0ODlkZjk1ZTU4MzU0OWI5MTlkZiIsICJmYW1pbHlfbmFtZSI6ICIiLCAiYXVkIjogImxtcy1rZXkiLCAiaXNzIjogImh0dHA6Ly9sb2NhbGhvc3Q6MTgwMDAvb2F1dGgyIiwgImVtYWlsX3ZlcmlmaWVkIjogdHJ1ZSwgIm5hbWUiOiAiIiwgInZlcnNpb24iOiAiMS4xLjAiLCAiZ2l2ZW5fbmFtZSI6ICIiLCAiZXhwIjogMTYwMzk5NDYzMywgImlzX3Jlc3RyaWN0ZWQiOiBmYWxzZSwgImVtYWlsIjogImVkeEBleGFtcGxlLmNvbSJ9'
    jwt_signature = 'G1fxJFH1zPZjs_ZT1e_oj_qpE3-9Wb0N4QN_RooRlXVVNHuYnUUBVi5p6W_QAJdv2V2EJiwzaDk4nGuOo1iV0-p0aCawg9XASePxWdskyyW74elyX94rmEeLn4-u2wjw7GxOx8bcZk_vXPeERoiMyFdVuXoJL3O8H-gQmkxe81sG8iJ93QZ4l-4D4fPJltL5O_UT7POyjBPQG-TEsjIQd8wcn4zdcDdwjh3Td3HzBuiKeVc-U9_rJrtQcmi8uZ-eQJmeq0u7gI-Lg9H-3X_of_cXe8A7xnB0lRetx9I6DDoHdrnP83bkZ5PUnhSZo-u3zP1G54yGcCSzHUiw64pApA'

    jwt_token = '{}.{}.{}'.format(jwt_header, jwt_payload, jwt_signature)

    headers = {
        'Authorization': 'JWT {}'.format(jwt_token)
    }

    requests.post(migrate_url, files={'file': file}, headers=headers)
    ```
2. You should see something like this in console of your devstack:
    ```
    2020-10-29 17:35:35,157 INFO 6099 [openedx.core.djangoapps.user_api.completion.tasks] tasks.py:42 - Started progress migration. Items to process: 1
    2020-10-29 17:35:35,157 INFO 6099 [openedx.core.djangoapps.user_api.completion.tasks] tasks.py:99 - Started progress migration from "source@example.com" to "target1@example.com" for "course-v1:OpenCraft+Course1+C1" course
    2020-10-29 17:35:35,211 INFO 6099 [completion_aggregator.batch] batch.py:104 - Performing aggregation update for 1 user enrollments
    2020-10-29 17:35:35,212 INFO 6099 [completion_aggregator.tasks.aggregation_tasks] aggregation_tasks.py:62 - Updating aggregators in course-v1:OpenCraft+Course1+C1 for target1. Changed blocks: set([BlockUsageLocator(CourseLocator(u'OpenCraft', u'Course1', u'C1', None, None), u'about', u'short_description')])
    2020-10-29 17:35:35,232 INFO 6099 [completion_aggregator.batch] batch.py:126 - Finished aggregation update for 1 user enrollments
    2020-10-29 17:35:35,252 INFO 6099 [completion_aggregator.batch] batch.py:104 - Performing aggregation update for 1 user enrollments
    2020-10-29 17:35:35,253 INFO 6099 [completion_aggregator.tasks.aggregation_tasks] aggregation_tasks.py:62 - Updating aggregators in course-v1:OpenCraft+Course1+C1 for target1. Changed blocks: set([BlockUsageLocator(CourseLocator(u'OpenCraft', u'Course1', u'C1', None, None), u'about', u'overview')])
    2020-10-29 17:35:35,270 INFO 6099 [completion_aggregator.batch] batch.py:126 - Finished aggregation update for 1 user enrollments
    2020-10-29 17:35:35,272 INFO 6099 [openedx.core.djangoapps.user_api.completion.tasks] tasks.py:178 - User progress in "course-v1:OpenCraft+Course1+C1" course successfully migrated from "source@example.com" to "target1@example.com"
    2020-10-29 17:35:35,273 INFO 6099 [openedx.core.djangoapps.user_api.completion.tasks] tasks.py:93 - Email with users progress migration results sent to [u'devops@example.com']
    ```
3. You should see target user enrolled to the course and with the same submissions as first user.

**Author notes and concerns**: None

**Reviewers**
- [ ] @giovannicimolin 
- [ ] edX reviewer[s] TBD

### CODE REVIEW

File: openedx/core/djangoapps/user_api/completion/views.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            return Response(status=400)
+
+        # Extract list to be used in migration task
+        migrate_list = [
+            (row['course'], row['source_email'], row['dest_email']) for row in reader
+            if row.get('outcome') != OUTCOME_MIGRATED  # Ignore lines marked as migrated
+        ]
+
+        # Start background task to migrate progress for given users
+        migrate_progress.delay(
```
Comment (giovannicimolin) [inline openedx/core/djangoapps/user_api/completion/views.py:46]: Did you try this task running the workers externally? The default vagrant devstack behavior is to run celery on eager mode.

Comment (0x29a) [reply openedx/core/djangoapps/user_api/completion/views.py:55]: @giovannicimolin, I use docker-based devstack. Tried to set `CELERY_ALWAYS_EAGER` to `False`, restarted lms, but received this error after sending POST to view:
```
Traceback:  

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/exception.py" in inner
  41.             response = get_response(request)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _legacy_get_response
  249.             response = self._get_response(request)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _get_response
  187.                 response = self.process_exception_by_middleware(e, request)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py" in _get_response
  185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py" in inner
  185.                     return func(*args, **kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/csrf.py" in wrapped_view
  58.         return view_func(*args, **kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py" in view
  68.             return self.dispatch(request, *args, **kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py" in dispatch
  489.             response = self.handle_exception(exc)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py" in handle_exception
  449.             self.raise_uncaught_exception(exc)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py" in dispatch
  486.             response = handler(request, *args, **kwargs)

File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/user_api/completion/views.py" in post
  57.             [recepient_address] if recepient_address else None

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py" in delay
  461.         return self.apply_async(args, kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py" in apply_async
  573.             **dict(self._get_exec_options(), **options)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/base.py" in send_task
  354.                 reply_to=reply_to or self.oid, **options

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/amqp.py" in publish_task
  310.             **kwargs

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/messaging.py" in publish
  172.                        routing_key, mandatory, immediate, exchange, declare)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in _ensured
  470.                                            interval_max)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in ensure_connection
  382.                         interval_start, interval_step, interval_max, callback)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/utils/__init__.py" in retry_over_time
  246.             return fun(*args, **kwargs)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in connect
  250.         return self.connection

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in connection
  756.                 self._connection = self._establish_connection()

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/connection.py" in _establish_connection
  711.         conn = self.transport.establish_connection()

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/kombu/transport/pyamqp.py" in establish_connection
  116.         conn = self.Connection(**opts)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/connection.py" in __init__
  165.         self.transport = self.Transport(host, connect_timeout, ssl)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/connection.py" in Transport
  186.         return create_transport(host, connect_timeout, ssl)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/transport.py" in create_transport
  299.         return TCPTransport(host, connect_timeout)

File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/amqp/transport.py" in __init__
  95.             raise socket.error(last_err)

Exception Type: error at /api/user/v1/progress/migrate/
Exception Value: [Errno 111] Connection refused
```
I guess Celery worker should be started somehow.

File: openedx/core/djangoapps/user_api/completion/tasks.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            for submission in submissions:
+                submission.student_id = anonymous_id_for_user(target, course_key)
+                submission.save()
+
+            # Migrate StudentModule
+            for state in student_states:
+                state.student = target
+                state.save()
+
+    except Exception:
```
Comment (giovannicimolin) [inline openedx/core/djangoapps/user_api/completion/tasks.py:174]: @0x29a Can you be a little bit more transparent about the issue? 
```suggestion
    except Exception as exc:
        log.exception("Unexpected error while migrating user progress: {}".format(str(exc))  # or something similar?
```

File: openedx/core/djangoapps/user_api/completion/views.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    )
+
+    permission_classes = (
+        permissions.IsAuthenticated,
+        permissions.IsAdminUser,
+    )
+
+    def post(self, request):
+        """
+        POST /api/user/v1/completion/migrate/
```
Comment (giovannicimolin) [inline openedx/core/djangoapps/user_api/completion/views.py:35]: Can you improve this a bit and the required inputs? i.e. how should the request be and how should the CSV be?

**Review by giovannicimolin — APPROVED** [review]
:+1:

- [x] I tested this:
1. Checked out this branch on my solutions devstack.
2. Create a new test course ([course.zwme_g4p.tar.gz](https://github.com/edx-solutions/edx-platform/files/5478343/course.zwme_g4p.tar.gz)).
3. Enabled completion tracking.
4. Created a new user account, and completed all the course items with it. Checked that all units were marked as complete.
5. Created a new empty user.
6. Modified and ran the migration script in the ticket's description.
7. Checked that the old user was not enrolled anymore.
8. Checked that the new user was enrolled and that completion and responses were transferred over.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ NA
- [ ] Includes documentation: no. Will be provided in internal ticket and some comments in code :x:

@0x29a Great work. Seems to be working perfectly :smile: 
You can merge after addressing the nits in the code.

File: openedx/core/djangoapps/user_api/completion/tasks.py (hunk 2 above)
Comment (giovannicimolin) [reply openedx/core/djangoapps/user_api/completion/tasks.py:174]: This will help us debug issues later.

Comment (0x29a) [reply openedx/core/djangoapps/user_api/completion/tasks.py:174]: @giovannicimolin, `log.exception` itself fetches traceback and throws it into output. Consider this example:
```python
import logging

logger = logging.getLogger(__name__)

def blackbox():
  raise ZeroDivisionError("Hey!")

try:
  blackbox()
except:
  logger.exception("Unexpected error.")

print("The End.")

```

Here is output for this:
```
➜  ~ python3 blackbox.py
Unexpected error.
Traceback (most recent call last):
  File "blackbox.py", line 9, in <module>
    blackbox()
  File "blackbox.py", line 6, in blackbox
    raise ZeroDivisionError("Hey!")
ZeroDivisionError: Hey!
The End.
```

Is that what we want?

Comment (giovannicimolin) [reply openedx/core/djangoapps/user_api/completion/tasks.py:174]: Ah, nice! Thanks for the heads up on this one.

**Pull request merged**